elapid/evaluate.py

```diff
@@ -0,0 +1,189 @@
+    Generates habitat suitability intervals for the Boyce Index calculation.
+
+    Calculates intervals based on the provided range and either the number of bins or bin size.
+
+    Args:
+        nbins (int or list or np.ndarray, optional): Number of classes or a list of class thresholds. Defaults to 0.
+        bin_size (float or str, optional): Width of the bins. Defaults to 'default', which sets the width as 1/10th of the range.
+        range (tuple or list): Two elements representing the minimum and maximum values of habitat suitability. Default : [0, 1]
+
+    Returns:
+        np.ndarray: An array of intervals, each represented by a pair of lower and upper bounds.
+
+    Raises:
+        ValueError: If invalid values are provided for nbins or bin_size.
+    """
+    mini, maxi = range
+
+    if isinstance(bin_size, float):
+        nbins = (maxi - mini) / bin_size
+        if not nbins.is_integer():
+            warnings.warn(
+                f"bin_size has been adjusted to nearest appropriate size using ceil, as range/bin_size : {(maxi - mini)} / {bin_size} is not an integer.",
+                UserWarning,
+            )
+        nbins = math.ceil(nbins)
+    elif nbins == 0 and bin_size == "default":
+        nbins = 10
+
+    if isinstance(nbins, (list, np.ndarray)):
+        if len(nbins) == 1:
+            raise ValueError("Invalid nbins value. len(nbins) must be > 1")
+        nbins.sort()
+        intervals = np.column_stack((nbins[:-1], nbins[1:]))
+    elif nbins > 1:
+        boundary = np.linspace(mini, maxi, num=nbins + 1)
+        intervals = np.column_stack((boundary[:-1], boundary[1:]))
+    else:
+        raise ValueError("Invalid nbins value. nbins > 1")
+
+    return intervals
+
+
+def boyce_index(yobs: np.ndarray, ypred: np.ndarray, interval: Union[Tuple[float, float], List[float]]) -> float:
+    """
+    Calculates the Boyce index for a given interval.
+
+    Uses the convention as defined in Hirzel et al. 2006 to compute the ratio of observed to expected frequencies.
+
+    Args:
+        yobs (np.ndarray): Suitability values at observed locations (e.g., predictions at presence points).
+        ypred (np.ndarray): Suitability values at random locations (e.g., predictions at background points).
+        interval (tuple or list): Two elements representing the lower and upper bounds of the interval (i.e., habitat suitability).
+
+    Returns:
+        float: The ratio of observed to expected frequencies for the given interval.
+
+    """
+    yobs_bin = (yobs >= interval[0]) & (yobs <= interval[1])
+    ypred_bin = (ypred >= interval[0]) & (ypred <= interval[1])
+
+    pi = np.sum(yobs_bin) / len(yobs_bin)
+    ei = np.sum(ypred_bin) / len(ypred_bin)
+
+    if ei == 0:
+        fi = np.nan  # Avoid division by zero
+    else:
+        fi = pi / ei
+
+    return fi
+
+
+def continuous_boyce_index(
+    yobs: Union[np.ndarray, pd.Series, gpd.GeoSeries],
+    ypred: Union[np.ndarray, pd.Series, gpd.GeoSeries],
```

Comment:
I think I had misunderstood the original implementation, and it looks like these are both actually predicted values, just predictions at different locations (ypred at presence and background sites).

I might prefer we rename these variables to be a bit more clear (like `ypred_observed` and `ypred_background`).